tests/test_evaluator.py

```diff
@@ -77,6 +78,51 @@ class TestEvaluator(TestCase):
     def test_wrong_task(self):
         self.assertRaises(KeyError, evaluator, "bad_task")
 
+    def test_device_placement(self):
+        orig_import = __import__
+
+        pt_mock = mock.Mock()
+        tf_mock = mock.Mock()
+
+        # mock import of torch and tensorflow
+        def import_pt_tf_mock(name, *args):
+            if name == "torch":
+                if pt_available:
+                    return pt_mock
+                else:
+                    raise ImportError
+            if name == "tensorflow":
+                if tf_available:
+                    return tf_mock
+                else:
+                    raise ImportError
+            return orig_import(name, *args)
+
+        with mock.patch("builtins.__import__", side_effect=import_pt_tf_mock):
+            # neither pt or tf are available
+            pt_available = False
+            tf_available = False
+            self.assertEqual(Evaluator._infer_device(), -1)
```

Comment:
Interesting, I wonder how this test passes since my understanding is that `device` is never explicitly defined when neither `torch` or `tf` are installed 🤔

Reply:
If neither tf or torch are installed then the second `except` sets it to `-1`:
https://github.com/huggingface/evaluate/blob/b48e48ce7287a875924953e2baff8bac426c644d/src/evaluate/evaluator/base.py#L149-L150